src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -80,6 +80,7 @@ module Crystal
     @block_context : Block?
     @file_module : FileModule?
     @while_vars : MetaVars?
+    @while_ensure_stack : Array(Symbol)
```

Comment:
I'm not fully convinced about the name.
Maybe should be `block_kind_stack` or something like that. Since it does not only trace while and ensure.

But, check the next comment, I think we don't even need an Array here.